proposals/p2015.md

```diff
@@ -0,0 +1,270 @@
+
+We want to establish a syntax for fixed-size scalar number types. These types
+include the two's complement signed integer, the unsigned integer, and the
+floating-point number.
+
+As these types are pervasive throughout the language, our goal here is to align
+on a terse, convenient, yet understandable, and ergonomic syntax to the author.
+
+## Background
+
+For developer convenience, names are given to number types that map to native
+machine register widths. These sizes typically include 8-bit, 16-bit, 32-bit,
+64-bit, and, more recently, 128-bit widths.
+
+For example, in [C++11+](https://en.cppreference.com/w/cpp/types/integer),
+integer types such as `int8_t` (8-bit two's complement signed integer type) and
+`uint16_t` (16-bit unsigned integer type) exist, among similar types for 32- and
+64-bit values. Correspondingly, you have the `i8` and `u16`
+([among others](https://doc.rust-lang.org/book/ch03-02-data-types.html#scalar-types))
+scalar integer types in Rust. And in Swift, the `Int8` and `UInt16`
+([among others](https://developer.apple.com/documentation/swift/uint8)) integer
+value types.
+
+In each case, the intent is to provide a clear and pragmatic syntax.
+
+Additional discussion around this proposal's background can be found in
+[#543](https://github.com/carbon-language/carbon-lang/issues/543).
+
+## Proposal
+
+We introduce a simple keyword-like syntax of `iN`, `uN`, and `fN` for two's
+complement integers, unsigned integers, and floating-point numbers,
+respectively. Where `N` can be a positive multiple of 8, including the common
+power-of-two sizes (for example, `N = 8, 16, 32`). This structure follows the
+successful precedent set by Rust and LLVM development communities and
+potentially saves 40% or more on characters required compared to other options
+such as `IntN` (for example, `i16` versus `Int16`). While bit sizes greater than
+128-bits will be well-supported, some operations like division will not be
+available on these large sizes.
+
+### Non-goals
+
+-   This does not address any considerations around the `bool` type
+-   This does not provide a formal plan for the shape or mapping of the
+    underlying types
+    ([#767 comments](https://github.com/carbon-language/carbon-lang/issues/767#issuecomment-1214153375))
+-   This does not prescribe an official grammar for parsing these types
+-   This proposal does not address other, non-multiple of 8 bit sizes, such as
+    those used in a bit field
+
+## Details
+
+### Syntax
+
+The syntax for a two's complement signed integer, the unsigned integer, and the
+floating-point number corresponds to a lowercase 'i', 'u', or 'f' character,
+respectively, indicating the type followed by a numeric value specifying the
+width.
+
+As a regular expression, this can be illustrated as:
+
+```re
+([iuf])([1-9][0-9]*)
+```
+
+Capture group 1 indicates either an 'i' for a two's complement signed integer
+type, a 'u' for an unsigned integer type, or an 'f' for a floating-point number
```

Comment:
It is good to specify that these are following the standard. It is also good to say this is a "binary" floating-point type, since that standard defines both binary and decimal floating-point formats.

```suggestion
type, a 'u' for an unsigned integer type, or an 'f' for an
[IEEE-754](https://en.wikipedia.org/wiki/IEEE_754) binary floating-point number
```